From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Martijn Muijsers <martijnmuijsers@live.nl>
Date: Mon, 7 Aug 2023 21:35:57 +0200
Subject: [PATCH] Make ender dragon respawn attempt after placing end crystals
 configurable

The original Pufferfish description has been included below for reference,
but please note that since Paper 1.20.1, there is no need to disable end crystal
spawning for performance: Paper already contains performance
enhancements that make the end crystal checks very fast.

See Paper#9611 for more information. This patch now only
functions as a gameplay mechanism tweak.

License: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)
Gale - https://galemc.org

This patch is based on the following patch:
"Option to disable end crystal dragon respawning"
By: Kevin Raneri <kevin.raneri@gmail.com>
As part of: Pufferfish (https://github.com/pufferfish-gg/Pufferfish)
Licensed under: GPL-3.0 (https://www.gnu.org/licenses/gpl-3.0.html)

* Pufferfish description *

On servers that expect end crystal fights in the end dimension,
disabling this will prevent the server from performing an expensive
search to attempt respawning the ender dragon whenever a player places
an end crystal.

diff --git a/net/minecraft/world/item/EndCrystalItem.java b/net/minecraft/world/item/EndCrystalItem.java
index 6e4a3c9c090a68bf5785a01d815c275411859c7b..0db35674726e3039f2c20aed1993d6d3843f8d29 100644
--- a/net/minecraft/world/item/EndCrystalItem.java
+++ b/net/minecraft/world/item/EndCrystalItem.java
@@ -49,10 +49,12 @@ public class EndCrystalItem extends Item {
                         // CraftBukkit end
                         level.addFreshEntity(endCrystal);
                         level.gameEvent(context.getPlayer(), GameEvent.ENTITY_PLACE, blockPos);
+                        if (level.galeConfig().gameplayMechanics.tryRespawnEnderDragonAfterEndCrystalPlace) { // Gale - Pufferfish - make ender dragon respawn attempt after placing end crystals configurable
                         EndDragonFight dragonFight = ((ServerLevel)level).getDragonFight();
                         if (dragonFight != null) {
                             dragonFight.tryRespawn(aboveBlockPos); // Paper - Perf: Do crystal-portal proximity check before entity lookup
                         }
+                        } // Gale - Pufferfish - make ender dragon respawn attempt after placing end crystals configurable
                     }
 
                     context.getItemInHand().shrink(1);
